drivers/builtin/src/nist_kw.c

```diff
@@ -318,80 +299,111 @@ static int unwrap(mbedtls_nist_kw_context *ctx,
  * KW-AD as defined in SP 800-38F section 6.2
  * KWP-AD as defined in SP 800-38F section 6.3
  */
-int mbedtls_nist_kw_unwrap(mbedtls_nist_kw_context *ctx,
-                           mbedtls_nist_kw_mode_t mode,
-                           const unsigned char *input, size_t in_len,
-                           unsigned char *output, size_t *out_len, size_t out_size)
+psa_status_t mbedtls_nist_kw_unwrap(mbedtls_svc_key_id_t key,
+                                    mbedtls_nist_kw_mode_t mode,
+                                    const unsigned char *input, size_t input_length,
+                                    unsigned char *output, size_t output_size,
+                                    size_t *output_length)
 {
-    int ret = 0;
-    size_t olen;
+    psa_status_t ret = 0;
     unsigned char A[KW_SEMIBLOCK_LENGTH];
     int diff;
+    size_t part_length, padlen = 0, Plen;
+    psa_cipher_operation_t unwrap_operation = PSA_CIPHER_OPERATION_INIT;
+    psa_key_attributes_t attributes;
 
-    *out_len = 0;
-    if (out_size < in_len - KW_SEMIBLOCK_LENGTH) {
-        return MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA;
+    ret = psa_get_key_attributes(key, &attributes);
+    if (ret != PSA_SUCCESS) {
+        goto cleanup;
+    }
+
+    /*
+     * PSA API are able to check the algorithm and usage as this is specified
+     * when the key is imported in the parameters. It is however unable to verify the
+     * key type as this is a requirement of the NIST KW specification, therefore add
+     * a check here to ensure the correct key type is used.
+     */
+    if (psa_get_key_type(&attributes) != PSA_KEY_TYPE_AES) {
+        ret = PSA_ERROR_NOT_PERMITTED;
+        goto cleanup;
+    }
+
+    ret = psa_cipher_decrypt_setup(&unwrap_operation, key, PSA_ALG_ECB_NO_PADDING);
+    if (ret != PSA_SUCCESS) {
+        goto cleanup;
+    }
+    *output_length = 0;
+    if (output_size < input_length - KW_SEMIBLOCK_LENGTH) {
+        ret = PSA_ERROR_INVALID_ARGUMENT;
```

Comment:
As [previously noted](https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/180#discussion_r2047281373), when the output buffer is too small, the error code should be `PSA_ERROR_BUFFER_TOO_SMALL`. The error `INVALID_ARGUMENT` conveys that some data is bad (whether it's an unknown value in an enum-like integer, a known value that isn't in an appropriate category, an integer out of range, a badly formatted string, etc.). If an error can be solved by allocating more memory for a buffer, without changing any data, that error should be conveyed by `BUFFER_TOO_SMALL`.